### PR TITLE
Add operators support for `$addFields`

### LINKED ISSUE #2679: Add operators support for `$addFields`
After #2678.

### What should be done?

Add operators support to `$addFields`.

#### Supported stages support operators
- `$group` stage `_id` expression supports operators
example: `{_id : { $dateToString: { format: "%Y-%m-%d", date: "$date" } } }` from [manual](https://www.mongodb.com/docs/manual/reference/operator/aggregation/group/#group-by-day-of-the-year)
https://github.com/FerretDB/FerretDB/blob/0992e5f8a410e59041a62a12f666eacf6f3bd27b/internal/handlers/common/aggregations/stages/group.go#L165 
- `$project` stage supports operators
example: `{$project: {isbn: {prefix: { $substr: [ "$isbn", 0, 3 ] } } } }` from [manual](https://www.mongodb.com/docs/manual/reference/operator/aggregation/project/#include-computed-fields)
https://github.com/FerretDB/FerretDB/blob/0992e5f8a410e59041a62a12f666eacf6f3bd27b/internal/handlers/common/aggregations/stages/projection/projection.go#L361
- query projection supports [operators](https://www.mongodb.com/docs/manual/reference/method/db.collection.find/#projection)
same example as `$project` works on projection
https://github.com/FerretDB/FerretDB/blob/0992e5f8a410e59041a62a12f666eacf6f3bd27b/internal/handlers/common/projection.go#L371

#### Out of scope
- `$match` is handled by https://github.com/FerretDB/FerretDB/issues/414
- `$group` stage recursive operator is handled by https://github.com/FerretDB/FerretDB/issues/2694

### Definition of Done

- all stages updated;
- compatibility tests for every operator in every stage added;
- spot refactorings done;
- find issue number in the code, skipped tests with this issue number passes.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #2679.

## Readiness checklist

- [ ] I added/updated unit tests.
- [x] I added/updated integration/compatibility tests.
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

**Comment by chilagrow** [comment]
Will there be another PR? The issue mentions the support of operator for common/projection.go but I'm not seeing the change related to that :thinking:

File: internal/handlers/common/add_fields_iterator.go [hunk 1]
```diff
@@ -65,6 +87,58 @@ func (iter *addFieldsIterator) Close() {
...
+		return nil
+	}
+
+	var opErr operators.OperatorError
+
+	if !errors.As(err, &opErr) {
+		return err
+	}
+
+	switch opErr.Code() {
```
Comment (chilagrow) [inline internal/handlers/common/add_fields_iterator.go:103]: Please add tests for checking error message for each of new errors please :pray:
We should be able to check the message for `ErrTooManyFields`, `ErrArgsInvalidLen` `ErrInvalidExpression`  and `ErrInvalidNestedExpression`.

File: integration/aggregate_documents_compat_test.go [hunk 2]
```diff
@@ -1618,20 +1618,20 @@ func TestAggregateCompatProject(t *testing.T) {
 		"TypeRecursive": {
 			pipeline: bson.A{
 				bson.D{{"$sort", bson.D{{"_id", -1}}}},
-				bson.D{{"$project", bson.D{{"type", bson.D{{"$type", bson.D{{"$type", bson.D{{"$type", "$v"}}}}}}}}}},
```
Comment (chilagrow) [inline integration/aggregate_documents_compat_test.go:1621]: Does this not work? Or is this not a meaningful test case? From my understand 2 level of nesting and 3 level of nesting just gives the same response :hugs:

File: integration/aggregate_documents_compat_test.go [hunk 3]
```diff
@@ -1806,6 +1812,127 @@ func TestAggregateCompatAddFields(t *testing.T) {
...
+				bson.D{{"$addFields", bson.D{{"type", bson.D{{"$type", bson.D{{"foo", "bar"}}}}}}}},
+			},
+		},
+		"TypeEmpty": {
+			pipeline: bson.A{
+				bson.D{{"$sort", bson.D{{"_id", -1}}}},
+				bson.D{{"$addFields", bson.D{{"type", bson.D{{"$type", bson.D{}}}}}}},
+			},
+		},
+		"MultipleOperators": {
```
Comment (chilagrow) [inline integration/aggregate_documents_compat_test.go:1884]: Great test!

**Review by AlekSi — COMMENTED** [review]
At the very least, there is a merge conflict

File: integration/aggregate_documents_compat_test.go (hunk 2 above)
Comment (noisersup) [reply integration/aggregate_documents_compat_test.go:1621]: It worked before and it works currently, the issue here is that my goal here was to check for recursion and not for double recursion. Also 2level of nesting returns the same results as 3level

File: internal/handlers/common/add_fields_iterator.go (hunk 1 above)
Comment (noisersup) [reply internal/handlers/common/add_fields_iterator.go:103]: The only error not tested from this list yet is ErrInvalidExpression, adding test for it.

File: internal/handlers/common/aggregations/stages/projection/projection.go [hunk 4]
```diff
@@ -643,7 +652,15 @@ func processOperatorError(err error) error {
...
-	case operators.ErrNoOperator, operators.ErrWrongType:
+	case operators.ErrNoOperator:
+		return commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrEmptySubProject,
+			"Invalid $project :: caused by :: An empty sub-projection is not a valid value."+
+				" Found empty object at path",
+			"$project (stage)",
+		)
+
+	case operators.ErrWrongType:
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/projection/projection.go:689]: It shouldn't be reachable, although maybe it's not needed

File: internal/handlers/common/add_fields_iterator.go (hunk 1 above)
Comment (chilagrow) [reply internal/handlers/common/add_fields_iterator.go:103]: Yep thanks!

File: internal/handlers/common/aggregations/stages/projection/projection.go [hunk 5]
```diff
@@ -232,6 +248,22 @@ func ProjectDocument(doc, projection *types.Document, inclusion bool) (*types.Do
...
 
+			if !operators.IsOperator(idValue) {
+				if idValue.Len() == 0 {
+					return nil, commonerrors.NewCommandErrorMsgWithArgument(
+						commonerrors.ErrEmptySubProject,
+						"Invalid $project :: caused by :: An empty sub-projection is not a valid value."+
+							" Found empty object at path",
+						"$project (stage)",
+					)
+				}
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/projection/projection.go:259]: Great catch finding this. Could you also test this please?

File: internal/handlers/common/aggregations/stages/projection/projection.go [hunk 6]
```diff
@@ -242,6 +274,7 @@ func ProjectDocument(doc, projection *types.Document, inclusion bool) (*types.Do
 				return nil, err
 			}
 
+			set = true
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/projection/projection.go:277]: This is no covered by tests, I think projection doesn't have any operator tests. Also needs test for recursive operator tests for $project to go with these line changes.

File: internal/handlers/common/aggregations/stages/projection/projection.go [hunk 7]
```diff
@@ -312,6 +345,21 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
...
 
+			if !operators.IsOperator(value) {
+				if value.Len() == 0 {
+					return nil, commonerrors.NewCommandErrorMsgWithArgument(
+						commonerrors.ErrEmptySubProject,
+						"Invalid $project :: caused by :: An empty sub-projection is not a valid value."+
+							" Found empty object at path",
+						"$project (stage)",
+					)
+				}
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/projection/projection.go:356]: Also great detecting this. Codecov thinks it's missing test for it. Could you add it too please :pray:

**Comment by chilagrow** [comment]
:wave: @noisersup, this PR does not address what I analysed on the issue, maybe there is a confusion :thinking:  The issue was about supporting `$group` `$project` and query projection to support recursive operator. At the time of analysis, `$addField` and `$set` were not implemented so that's why only 3 of them support operators. Is there a confusion in the issue where it says *Supported stages support operators* in the issue please? If we want to handle `$addField` in this issue, that's great but could you create issues for remaining `$group`, `$project` and query projection and `$set` too please?  Maybe you've already created them :thinking:

**Comment by noisersup** [comment]
> The issue was about supporting $group $project and query projection to support recursive operator. 

I believe that from the description and title it seems that we need to support operators in more stages than only in `$project`. As we want to cut the scope for this issue I've created another issue that covers all of remaining operators -> #2881

File: internal/handlers/common/aggregations/stages/projection/projection.go (hunk 5 above)
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/projection/projection.go:259]: It's not possible as it's catched on validation.

File: internal/handlers/common/aggregations/stages/projection/projection.go (hunk 7 above)
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/projection/projection.go:356]: The same thing as above, we have a test for this but it's catched on validation

Comment (noisersup) [reply internal/handlers/common/aggregations/stages/projection/projection.go]: After DS we decided to remove both unreachable code as everything is already catched on validation

**Review by rumyantseva — APPROVED** [review]
Looks good to me, maybe Chi can spot something as she analyzed the original task, but to me the code looks consistent with the rest of aggregation.

**Review by chilagrow — APPROVED** [review]
Thanks for this!

File: internal/handlers/common/aggregations/stages/projection/projection.go (hunk 7 above)
Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/projection/projection.go]: Good idea of removing unreachable code :+1:

File: internal/handlers/common/aggregations/stages/projection/projection.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/projection/projection.go]: Thanks for removing this, good idea :+1:

File: internal/handlers/common/aggregations/stages/projection/projection.go (hunk 6 above)
Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/projection/projection.go:268]: Now covered

**Pull request merged**